src/utils/image-processor.js

```diff
@@ -91,90 +92,73 @@ const shrinkFile = function(file, settings) {
+    if (img.width * step * img.height * step < settings.size) {
+      return df.reject('not required')
     }
-    return defer(function() {
-      var canvas
-      sW = Math.round(sW * step)
-      sH = Math.round(sH * step)
-      if (sW * step < w) {
-        sW = w
-        sH = h
-      }
-      if (sW * sH > maxSquare) {
-        sW = Math.floor(Math.sqrt(maxSquare * ratio))
-        sH = Math.floor(maxSquare / Math.sqrt(maxSquare * ratio))
-      }
-      if (sW > maxSize) {
-        sW = maxSize
-        sH = Math.round(sW / ratio)
-      }
-      if (sH > maxSize) {
-        sH = maxSize
-        sW = Math.round(ratio * sH)
+    const originalW = img.width
+    let sW = originalW
+    let sH = img.height
+    const ratio = sW / sH
+    const w = Math.floor(Math.sqrt(settings.size * ratio))
+    const h = Math.floor(settings.size / Math.sqrt(settings.size * ratio))
+
+    const run = function() {
+      if (sW <= w) {
+        df.resolve(img)
+        return
       }
-      canvas = document.createElement('canvas')
-      canvas.width = sW
-      canvas.height = sH
-      canvas.getContext('2d').drawImage(img, 0, 0, sW, sH)
-      img.src = '//:0' // for image
-      img.width = img.height = 1 // for canvas
-      img = canvas
-      df.notify((originalW - sW) / (originalW - w))
-      return run()
-    })
-  }
+      return defer(function() {
+        sW = Math.round(sW * step)
+        sH = Math.round(sH * step)
+        if (sW * step < w) {
+          sW = w
+          sH = h
+        }
+        if (sW * sH > maxSquare) {
+          sW = Math.floor(Math.sqrt(maxSquare * ratio))
+          sH = Math.floor(maxSquare / Math.sqrt(maxSquare * ratio))
+        }
+        if (sW > maxSize) {
+          sW = maxSize
+          sH = Math.round(sW / ratio)
+        }
+        if (sH > maxSize) {
+          sH = maxSize
+          sW = Math.round(ratio * sH)
+        }
+        const canvas = document.createElement('canvas')
+        const cx = canvas.getContext('2d')
+        if ('imageSmoothingQuality' in cx) {
+          cx.imageSmoothingQuality = 'high'
+        }
+        canvas.width = sW
+        canvas.height = sH
+        cx.drawImage(img, 0, 0, sW, sH)
 
-  runNative = function() {
-    var canvas, cx
-    canvas = document.createElement('canvas')
-    canvas.width = w
-    canvas.height = h
-    cx = canvas.getContext('2d')
-    cx.imageSmoothingQuality = 'high'
-    cx.drawImage(img, 0, 0, w, h)
-    img.src = '//:0' // for image
-    img.width = img.height = 1 // for canvas
-    return df.resolve(canvas)
-  }
+        img.src = '//:0' // for image
+        img.width = img.height = 1 // for canvas
+        img = canvas
 
-  cx = document.createElement('canvas').getContext('2d')
+        df.notify((originalW - sW) / (originalW - w))
+
+        return run()
+      })
+    }
+
+    if (maxSize > 0 && maxSquare > 0) {
+      run()
+    } else {
+      return df.reject('not supported')
+    }
+  })
```

Comment:
Another advantage of having separate `limitedCanvasSize` is we can reject requests inside it if requested size is far lager than browser capabilities. In this case we need proper `.fail()` handling here.